File: rust_projects/day10_review.rs
// Day 10: Comprehensive Review & Practice
// Topics: Mixed problem sets, pattern recognition, complexity analysis

use std::collections::{HashMap, HashSet, VecDeque, BinaryHeap};
use std::cmp::Reverse;

// ============================================================================
// PATTERN 1: Arrays & Hashing - Two Sum
// ============================================================================
fn two_sum(nums: &[i32], target: i32) -> Vec<usize> {
    let mut seen = HashMap::new();
    for (i, &num) in nums.iter().enumerate() {
        let complement = target - num;
        if let Some(&prev_idx) = seen.get(&complement) {
            return vec![prev_idx, i];
        }
        seen.insert(num, i);
    }
    vec![]
}

fn three_sum(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
    nums.sort();
    let mut result = Vec::new();
    let n = nums.len();
    
    for i in 0..n.saturating_sub(2) {
        if i > 0 && nums[i] == nums[i - 1] {
            continue;
        }
        
        let mut left = i + 1;
        let mut right = n - 1;
        
        while left < right {
            let total = nums[i] + nums[left] + nums[right];
            
            if total == 0 {
                result.push(vec![nums[i], nums[left], nums[right]]);
                while left < right && nums[left] == nums[left + 1] {
                    left += 1;
                }
                while left < right && nums[right] == nums[right - 1] {
                    right -= 1;
                }
                left += 1;
                right -= 1;
            } else if total < 0 {
                left += 1;
            } else {
                right -= 1;
            }
        }
    }
    
    result
}

fn max_area(height: &[i32]) -> i32 {
    let mut left = 0;
    let mut right = height.len() - 1;
    let mut max_area = 0;
    
    while left < right {
        let width = (right - left) as i32;
        let h = height[left].min(height[right]);
        max_area = max_area.max(width * h);
        
        if height[left] < height[right] {
            left += 1;
        } else {
            right -= 1;
        }
    }
    
    max_area
}

fn group_anagrams(strs: &[String]) -> Vec<Vec<String>> {
    let mut anagram_groups: HashMap<String, Vec<String>> = HashMap::new();
    
    for s in strs {
        let mut key = s.clone();
        let mut chars: Vec<char> = key.chars().collect();
        chars.sort_unstable();
        key = chars.into_iter().collect();
        anagram_groups.entry(key).or_insert_with(Vec::new).push(s.clone());
    }
    
    anagram_groups.into_values().collect()
}

// ============================================================================
// PATTERN 2: String Manipulation
// ============================================================================
fn length_of_longest_substring(s: &str) -> usize {
    let chars: Vec<char> = s.chars().collect();
    let mut char_set = HashSet::new();
    let mut left = 0;
    let mut max_length = 0;
    
    for right in 0..chars.len() {
        while char_set.contains(&chars[right]) {
            char_set.remove(&chars[left]);
            left += 1;
        }
        char_set.insert(chars[right]);
        max_length = max_length.max(right - left + 1);
    }
    
    max_length
}

fn count_palindromic_substrings(s: &str) -> usize {
    let chars: Vec<char> = s.chars().collect();
    let mut count = 0;
    
    fn expand(chars: &[char], left: usize, right: usize, count: &mut usize) {
        let mut l = left;
        let mut r = right;
        while l < chars.len() && r < chars.len() && chars[l] == chars[r] {
            *count += 1;
            if l == 0 { break; }
            l -= 1;
            r += 1;
        }
    }
    
    for i in 0..chars.len() {
        expand(&chars, i, i, &mut count);
        if i + 1 < chars.len() {
            expand(&chars, i, i + 1, &mut count);
        }
    }
    
    count
}

fn str_str(haystack: &str, needle: &str) -> i32 {
    if needle.is_empty() {
        return 0;
    }
    if needle.len() > haystack.len() {
        return -1;
    }
    
    for i in 0..=(haystack.len() - needle.len()) {
        if &haystack[i..i + needle.len()] == needle {
            return i as i32;
        }
    }
    
    -1
}

// ============================================================================
// PATTERN 3: Stack & Queue
// ============================================================================
struct MinStack {
    stack: Vec<i32>,
    min_stack: Vec<i32>,
}

impl MinStack {
    fn new() -> Self {
        MinStack {
            stack: Vec::new(),
            min_stack: Vec::new(),
        }
    }
    
    fn push(&mut self, val: i32) {
        self.stack.push(val);
        if self.min_stack.is_empty() || val <= *self.min_stack.last().unwrap() {
            self.min_stack.push(val);
        }
    }
    
    fn pop(&mut self) {
        if let Some(val) = self.stack.pop() {
            if Some(&val) == self.min_stack.last() {
                self.min_stack.pop();
            }
        }
    }
    
    fn top(&self) -> Option<i32> {
        self.stack.last().copied()
    }
    
    fn get_min(&self) -> Option<i32> {
        self.min_stack.last().copied()
    }
}

fn eval_rpn(tokens: Vec<String>) -> i32 {
    let mut stack = Vec::new();
    
    for token in tokens {
        match token.as_str() {
            "+" => {
                let b = stack.pop().unwrap();
                let a = stack.pop().unwrap();
                stack.push(a + b);
            }
            "-" => {
                let b = stack.pop().unwrap();
                let a = stack.pop().unwrap();
                stack.push(a - b);
            }
            "*" => {
                let b = stack.pop().unwrap();
                let a = stack.pop().unwrap();
                stack.push(a * b);
            }
            "/" => {
                let b = stack.pop().unwrap();
                let a = stack.pop().unwrap();
                stack.push(a / b);
            }
            _ => {
                stack.push(token.parse().unwrap());
            }
        }
    }
    
    stack[0]
}

fn daily_temperatures(temperatures: &[i32]) -> Vec<i32> {
    let mut result = vec![0; temperatures.len()];
    let mut stack: Vec<usize> = Vec::new();
    
    for (i, &temp) in temperatures.iter().enumerate() {
        while let Some(&prev_idx) = stack.last() {
            if temperatures[prev_idx] < temp {
                stack.pop();
                result[prev_idx] = (i - prev_idx) as i32;
            } else {
                break;
            }
        }
        stack.push(i);
    }
    
    result
}

// ============================================================================
// PATTERN 4: Linked List
// ============================================================================
#[derive(PartialEq, Eq, Clone, Debug)]
struct ListNode {
    val: i32,
    next: Option<Box<ListNode>>,
}

impl ListNode {
    fn new(val: i32) -> Self {
        ListNode { val, next: None }
    }
}

fn merge_two_lists(
    l1: Option<Box<ListNode>>,
    l2: Option<Box<ListNode>>,
) -> Option<Box<ListNode>> {
    match (l1, l2) {
        (None, None) => None,
        (Some(node), None) | (None, Some(node)) => Some(node),
        (Some(mut n1), Some(mut n2)) => {
            if n1.val < n2.val {
                n1.next = merge_two_lists(n1.next, Some(n2));
                Some(n1)
            } else {
                n2.next = merge_two_lists(Some(n1), n2.next);
                Some(n2)
            }
        }
    }
}

fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
    let mut prev = None;
    let mut curr = head;
    
    while let Some(mut node) = curr {
        curr = node.next.take();
        node.next = prev;
        prev = Some(node);
    }
    
    prev
}

// ============================================================================
// PATTERN 5: Trees
// ============================================================================
#[derive(Debug, PartialEq, Eq, Clone)]
struct TreeNode {
    val: i32,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

impl TreeNode {
    fn new(val: i32) -> Self {
        TreeNode { val, left: None, right: None }
    }
}

fn max_depth(root: Option<Box<TreeNode>>) -> i32 {
    match root {
        None => 0,
        Some(node) => 1 + max_depth(node.left).max(max_depth(node.right)),
    }
}

fn is_valid_bst(root: Option<Box<TreeNode>>) -> bool {
    fn validate(node: Option<&Box<TreeNode>>, low: i64, high: i64) -> bool {
        match node {
            None => true,
            Some(n) => {
                let val = n.val as i64;
                val > low && val < high
                    && validate(n.left.as_ref(), low, val)
                    && validate(n.right.as_ref(), val, high)
            }
        }
    }
    
    validate(root.as_ref(), i64::MIN, i64::MAX)
}

// ============================================================================
// PATTERN 6: Graphs
// ============================================================================
fn graph_bfs(graph: &HashMap<usize, Vec<usize>>, start: usize) -> Vec<usize> {
    let mut visited = HashSet::new();
    let mut result = Vec::new();
    let mut queue = VecDeque::new();
    
    queue.push_back(start);
    visited.insert(start);
    
    while let Some(node) = queue.pop_front() {
        result.push(node);
        
        if let Some(neighbors) = graph.get(&node) {
            for &neighbor in neighbors {
                if visited.insert(neighbor) {
                    queue.push_back(neighbor);
                }
            }
        }
    }
    
    result
}

fn dijkstra(graph: &HashMap<usize, Vec<(usize, i32)>>, start: usize) -> HashMap<usize, i32> {
    let mut distances = HashMap::new();
    
    for &node in graph.keys() {
        distances.insert(node, i32::MAX);
    }
    distances.insert(start, 0);
    
    let mut pq = BinaryHeap::new();
    pq.push(Reverse((0, start)));
    
    while let Some(Reverse((dist, node))) = pq.pop() {
        if dist > distances[&node] {
            continue;
        }
        
        if let Some(neighbors) = graph.get(&node) {
            for &(neighbor, weight) in neighbors {
                let new_dist = dist + weight;
                if new_dist < distances[&neighbor] {
                    distances.insert(neighbor, new_dist);
                    pq.push(Reverse((new_dist, neighbor)));
                }
            }
        }
    }
    
    distances
}

// ============================================================================
// TESTS
// ============================================================================
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_two_sum() {
        assert_eq!(two_sum(&[2, 7, 11, 15], 9), vec![0, 1]);
        assert_eq!(two_sum(&[3, 2, 4], 6), vec![1, 2]);
    }

    #[test]
    fn test_three_sum() {
        let result = three_sum(vec![-1, 0, 1, 2, -1, -4]);
        assert_eq!(result.len(), 2);
    }

    #[test]
    fn test_max_area() {
        assert_eq!(max_area(&[1, 8, 6, 2, 5, 4, 8, 3, 7]), 49);
    }

    #[test]
    fn test_group_anagrams() {
        let strs = vec!["eat".to_string(), "tea".to_string(), "tan".to_string(),
                        "ate".to_string(), "nat".to_string(), "bat".to_string()];
        let result = group_anagrams(&strs);
        assert_eq!(result.len(), 3);
    }

    #[test]
    fn test_length_of_longest_substring() {
        assert_eq!(length_of_longest_substring("abcabcbb"), 3);
        assert_eq!(length_of_longest_substring("bbbbb"), 1);
    }

    #[test]
    fn test_str_str() {
        assert_eq!(str_str("hello", "ll"), 2);
        assert_eq!(str_str("aaaaa", "bba"), -1);
        assert_eq!(str_str("", ""), 0);
    }

    #[test]
    fn test_min_stack() {
        let mut min_stack = MinStack::new();
        min_stack.push(-2);
        min_stack.push(0);
        min_stack.push(-3);
        assert_eq!(min_stack.get_min(), Some(-3));
        min_stack.pop();
        assert_eq!(min_stack.top(), Some(0));
        assert_eq!(min_stack.get_min(), Some(-2));
    }

    #[test]
    fn test_daily_temperatures() {
        let temps = vec![73, 74, 75, 71, 69, 72, 76, 73];
        let result = daily_temperatures(&temps);
        assert_eq!(result, vec![1, 1, 4, 2, 1, 1, 0, 0]);
    }

    #[test]
    fn test_reverse_list() {
        let mut head = Some(Box::new(ListNode::new(1)));
        head.as_mut().unwrap().next = Some(Box::new(ListNode::new(2)));
        head.as_mut().unwrap().next.as_mut().unwrap().next = Some(Box::new(ListNode::new(3)));
        
        let reversed = reverse_list(head);
        assert_eq!(reversed.as_ref().unwrap().val, 3);
        assert_eq!(reversed.as_ref().unwrap().next.as_ref().unwrap().val, 2);
    }

    #[test]
    fn test_max_depth() {
        let root = Some(Box::new(TreeNode {
            val: 3,
            left: Some(Box::new(TreeNode::new(9))),
            right: Some(Box::new(TreeNode {
                val: 20,
                left: Some(Box::new(TreeNode::new(15))),
                right: Some(Box::new(TreeNode::new(7))),
            })),
        }));
        
        assert_eq!(max_depth(root), 3);
    }

    #[test]
    fn test_is_valid_bst() {
        let root = Some(Box::new(TreeNode {
            val: 2,
            left: Some(Box::new(TreeNode::new(1))),
            right: Some(Box::new(TreeNode::new(3))),
        }));
        
        assert!(is_valid_bst(root));
    }

    #[test]
    fn test_graph_bfs() {
        let mut graph = HashMap::new();
        graph.insert(0, vec![1, 2]);
        graph.insert(1, vec![2]);
        graph.insert(2, vec![0, 3]);
        graph.insert(3, vec![3]);
        
        let result = graph_bfs(&graph, 2);
        assert_eq!(result.len(), 4);
    }

    #[test]
    fn test_dijkstra() {
        let mut graph = HashMap::new();
        graph.insert(0, vec![(1, 4), (2, 1)]);
        graph.insert(1, vec![(3, 1)]);
        graph.insert(2, vec![(1, 2), (3, 5)]);
        graph.insert(3, vec![]);
        
        let distances = dijkstra(&graph, 0);
        assert_eq!(distances[&0], 0);
        assert_eq!(distances[&1], 3);
        assert_eq!(distances[&2], 1);
        assert_eq!(distances[&3], 4);
    }
}

File: references/python-projects/day10_review.py
"""
Day 10: Comprehensive Review & Practice
Topics: Mixed problem sets, pattern recognition, complexity analysis, problem decomposition
"""

from collections import defaultdict, deque, Counter
import heapq
from typing import List, Dict, Set, Tuple, Optional


# ============================================================================
# PATTERN 1: Arrays & Hashing - Two Sum
# ============================================================================
def two_sum(nums: List[int], target: int) -> List[int]:
    """
    Two Sum - O(n) time, O(n) space
    Pattern: Hash map complement lookup
    """
    seen = {}
    for i, num in enumerate(nums):
        complement = target - num
        if complement in seen:
            return [seen[complement], i]
        seen[num] = i
    return []


def three_sum(nums: List[int]) -> List[List[int]]:
    """
    3Sum - O(n^2) time, O(1) extra space (excluding output)
    Pattern: Sorting + Two pointers
    """
    nums.sort()
    result = []
    n = len(nums)
    
    for i in range(n - 2):
        if i > 0 and nums[i] == nums[i - 1]:
            continue
        
        left, right = i + 1, n - 1
        while left < right:
            total = nums[i] + nums[left] + nums[right]
            
            if total == 0:
                result.append([nums[i], nums[left], nums[right]])
                while left < right and nums[left] == nums[left + 1]:
                    left += 1
                while left < right and nums[right] == nums[right - 1]:
                    right -= 1
                left += 1
                right -= 1
            elif total < 0:
                left += 1
            else:
                right -= 1
    
    return result


def max_area(height: List[int]) -> int:
    """
    Container With Most Water - O(n) time, O(1) space
    Pattern: Two pointers (opposite ends)
    """
    left, right = 0, len(height) - 1
    max_area = 0
    
    while left < right:
        width = right - left
        h = min(height[left], height[right])
        max_area = max(max_area, width * h)
        
        if height[left] < height[right]:
            left += 1
        else:
            right -= 1
    
    return max_area


def group_anagrams(strs: List[str]) -> List[List[str]]:
    """
    Group Anagrams - O(n * k * log(k)) time, O(n * k) space
    Pattern: Hash map with sorted key
    """
    anagram_groups = defaultdict(list)
    
    for s in strs:
        key = ''.join(sorted(s))
        anagram_groups[key].append(s)
    
    return list(anagram_groups.values())


# ============================================================================
# PATTERN 2: String Manipulation
# ============================================================================
def length_of_longest_substring(s: str) -> int:
    """
    Longest Substring Without Repeating - O(n) time, O(min(n, m)) space
    Pattern: Sliding window with hash set
    """
    char_set = set()
    left = 0
    max_length = 0
    
    for right in range(len(s)):
        while s[right] in char_set:
            char_set.remove(s[left])
            left += 1
        char_set.add(s[right])
        max_length = max(max_length, right - left + 1)
    
    return max_length


def count_palindromic_substrings(s: str) -> int:
    """
    Count Palindromic Substrings - O(n^2) time, O(1) space
    Pattern: Expand around center
    """
    count = 0
    
    def expand_around_center(left: int, right: int):
        nonlocal count
        while left >= 0 and right < len(s) and s[left] == s[right]:
            count += 1
            left -= 1
            right += 1
    
    for i in range(len(s)):
        expand_around_center(i, i)
        expand_around_center(i, i + 1)
    
    return count


def str_str(haystack: str, needle: str) -> int:
    """
    Implement strStr() - O(n * m) naive, O(n + m) with KMP
    Pattern: String matching
    """
    if not needle:
        return 0
    if len(needle) > len(haystack):
        return -1
    
    for i in range(len(haystack) - len(needle) + 1):
        if haystack[i:i + len(needle)] == needle:
            return i
    
    return -1


# ============================================================================
# PATTERN 3: Stack & Queue
# ============================================================================
class MinStack:
    """
    Min Stack - O(1) time for all operations, O(n) space
    Pattern: Auxiliary stack for minimum tracking
    """
    def __init__(self):
        self.stack = []
        self.min_stack = []
    
    def push(self, val: int) -> None:
        self.stack.append(val)
        if not self.min_stack or val <= self.min_stack[-1]:
            self.min_stack.append(val)
    
    def pop(self) -> None:
        if self.stack:
            val = self.stack.pop()
            if val == self.min_stack[-1]:
                self.min_stack.pop()
    
    def top(self) -> int:
        return self.stack[-1] if self.stack else None
    
    def get_min(self) -> int:
        return self.min_stack[-1] if self.min_stack else None


def eval_rpn(tokens: List[str]) -> int:
    """
    Evaluate Reverse Polish Notation - O(n) time, O(n) space
    Pattern: Stack for expression evaluation
    """
    stack = []
    
    for token in tokens:
        if token in {'+', '-', '*', '/'}:
            b = stack.pop()
            a = stack.pop()
            
            if token == '+':
                stack.append(a + b)
            elif token == '-':
                stack.append(a - b)
            elif token == '*':
                stack.append(a * b)
            else:
                stack.append(int(a / b))
        else:
            stack.append(int(token))
    
    return stack[0]


def daily_temperatures(temperatures: List[int]) -> List[int]:
    """
    Daily Temperatures - O(n) time, O(n) space
    Pattern: Monotonic decreasing stack
    """
    result = [0] * len(temperatures)
    stack = []
    
    for i, temp in enumerate(temperatures):
        while stack and temperatures[stack[-1]] < temp:
            prev_idx = stack.pop()
            result[prev_idx] = i - prev_idx
        stack.append(i)
    
    return result


def num_islands(grid: List[List[str]]) -> int:
    """
    Number of Islands - O(m * n) time, O(min(m, n)) space
    Pattern: BFS/DFS on grid
    """
    if not grid:
        return 0
    
    rows, cols = len(grid), len(grid[0])
    count = 0
    
    def bfs(r: int, c: int):
        queue = deque([(r, c)])
        grid[r][c] = '0'
        
        while queue:
            curr_r, curr_c = queue.popleft()
            
            for dr, dc in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
                new_r, new_c = curr_r + dr, curr_c + dc
                
                if (0 <= new_r < rows and 0 <= new_c < cols and 
                    grid[new_r][new_c] == '1'):
                    grid[new_r][new_c] = '0'
                    queue.append((new_r, new_c))
    
    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == '1':
                count += 1
                bfs(r, c)
    
    return count


# ============================================================================
# PATTERN 4: Linked List
# ============================================================================
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def merge_two_lists(l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    """
    Merge Two Sorted Lists - O(n + m) time, O(1) space
    Pattern: Iterative merging with dummy node
    """
    dummy = ListNode()
    curr = dummy
    
    while l1 and l2:
        if l1.val < l2.val:
            curr.next = l1
            l1 = l1.next
        else:
            curr.next = l2
            l2 = l2.next
        curr = curr.next
    
    curr.next = l1 if l1 else l2
    return dummy.next


def reverse_list(head: Optional[ListNode]) -> Optional[ListNode]:
    """
    Reverse Linked List - O(n) time, O(1) space
    Pattern: Three-pointer iteration
    """
    prev = None
    curr = head
    
    while curr:
        next_node = curr.next
        curr.next = prev
        prev = curr
        curr = next_node
    
    return prev


def is_palindrome(head: Optional[ListNode]) -> bool:
    """
    Palindrome Linked List - O(n) time, O(1) space
    Pattern: Fast-slow pointers + reversal
    """
    if not head or not head.next:
        return True
    
    # Find middle
    slow = fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
    
    # Reverse second half
    prev = None
    while slow:
        next_node = slow.next
        slow.next = prev
        prev = slow
        slow = next_node
    
    # Compare halves
    left, right = head, prev
    while right:
        if left.val != right.val:
            return False
        left = left.next
        right = right.next
    
    return True


# ============================================================================
# PATTERN 5: Trees
# ============================================================================
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def level_order(root: Optional[TreeNode]) -> List[List[int]]:
    """
    Binary Tree Level Order Traversal - O(n) time, O(n) space
    Pattern: BFS with queue
    """
    if not root:
        return []
    
    result = []
    queue = deque([root])
    
    while queue:
        level_size = len(queue)
        level = []
        
        for _ in range(level_size):
            node = queue.popleft()
            level.append(node.val)
            
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        
        result.append(level)
    
    return result


def max_depth(root: Optional[TreeNode]) -> int:
    """
    Maximum Depth of Binary Tree - O(n) time, O(h) space
    Pattern: DFS recursion
    """
    if not root:
        return 0
    return 1 + max(max_depth(root.left), max_depth(root.right))


def is_valid_bst(root: Optional[TreeNode]) -> bool:
    """
    Validate BST - O(n) time, O(h) space
    Pattern: DFS with range validation
    """
    def validate(node: Optional[TreeNode], low: float, high: float) -> bool:
        if not node:
            return True
        
        if not (low < node.val < high):
            return False
        
        return (validate(node.left, low, node.val) and
                validate(node.right, node.val, high))
    
    return validate(root, float('-inf'), float('inf'))


def diameter_of_binary_tree(root: Optional[TreeNode]) -> int:
    """
    Diameter of Binary Tree - O(n) time, O(h) space
    Pattern: DFS with global max tracking
    """
    diameter = [0]
    
    def depth(node: Optional[TreeNode]) -> int:
        if not node:
            return 0
        
        left_depth = depth(node.left)
        right_depth = depth(node.right)
        
        diameter[0] = max(diameter[0], left_depth + right_depth)
        
        return 1 + max(left_depth, right_depth)
    
    depth(root)
    return diameter[0]


def is_balanced(root: Optional[TreeNode]) -> bool:
    """
    Balanced Binary Tree - O(n) time, O(h) space
    Pattern: DFS with balance checking
    """
    def check_balance(node: Optional[TreeNode]) -> Tuple[int, bool]:
        if not node:
            return 0, True
        
        left_height, left_balanced = check_balance(node.left)
        right_height, right_balanced = check_balance(node.right)
        
        balanced = (left_balanced and right_balanced and
                   abs(left_height - right_height) <= 1)
        
        return 1 + max(left_height, right_height), balanced
    
    _, result = check_balance(root)
    return result


# ============================================================================
# PATTERN 6: Graphs
# ============================================================================
def graph_bfs(graph: Dict[int, List[int]], start: int) -> List[int]:
    """
    BFS on Graph - O(V + E) time, O(V) space
    Pattern: Queue-based traversal
    """
    visited = set()
    result = []
    queue = deque([start])
    visited.add(start)
    
    while queue:
        node = queue.popleft()
        result.append(node)
        
        for neighbor in graph[node]:
            if neighbor not in visited:
                visited.add(neighbor)
                queue.append(neighbor)
    
    return result


def topological_sort(num_courses: int, prerequisites: List[List[int]]) -> List[int]:
    """
    Topological Sort (Kahn's Algorithm) - O(V + E) time, O(V + E) space
    Pattern: BFS with in-degree tracking
    """
    in_degree = [0] * num_courses
    graph = defaultdict(list)
    
    for course, prereq in prerequisites:
        graph[prereq].append(course)
        in_degree[course] += 1
    
    queue = deque([i for i in range(num_courses) if in_degree[i] == 0])
    result = []
    
    while queue:
        node = queue.popleft()
        result.append(node)
        
        for neighbor in graph[node]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)
    
    return result if len(result) == num_courses else []


def dijkstra(graph: Dict[int, List[Tuple[int, int]]], start: int) -> Dict[int, int]:
    """
    Dijkstra's Algorithm - O((V + E) * log(V)) time, O(V + E) space
    Pattern: Priority queue for shortest path
    """
    distances = {node: float('inf') for node in graph}
    distances[start] = 0
    pq = [(0, start)]
    
    while pq:
        curr_dist, curr_node = heapq.heappop(pq)
        
        if curr_dist > distances[curr_node]:
            continue
        
        for neighbor, weight in graph[curr_node]:
            distance = curr_dist + weight
            
            if distance < distances[neighbor]:
                distances[neighbor] = distance
                heapq.heappush(pq, (distance, neighbor))
    
    return distances


# ============================================================================
# TESTING
# ============================================================================
def test_all():
    print("=" * 80)
    print("Day 10: Comprehensive Review & Practice - Running Tests")
    print("=" * 80)
    
    # Test Two Sum
    print("\n1. Two Sum")
    assert two_sum([2, 7, 11, 15], 9) == [0, 1]
    assert two_sum([3, 2, 4], 6) == [1, 2]
    print("   ✓ Two Sum tests passed")
    
    # Test 3Sum
    print("\n2. 3Sum")
    result = three_sum([-1, 0, 1, 2, -1, -4])
    assert sorted(result) == sorted([[-1, -1, 2], [-1, 0, 1]])
    print("   ✓ 3Sum tests passed")
    
    # Test Max Area
    print("\n3. Container With Most Water")
    assert max_area([1, 8, 6, 2, 5, 4, 8, 3, 7]) == 49
    print("   ✓ Max Area tests passed")
    
    # Test Group Anagrams
    print("\n4. Group Anagrams")
    result = group_anagrams(["eat", "tea", "tan", "ate", "nat", "bat"])
    assert len(result) == 3
    print("   ✓ Group Anagrams tests passed")
    
    # Test Longest Substring
    print("\n5. Longest Substring Without Repeating")
    assert length_of_longest_substring("abcabcbb") == 3
    assert length_of_longest_substring("bbbbb") == 1
    print("   ✓ Longest Substring tests passed")
    
    # Test Palindromic Substrings
    print("\n6. Count Palindromic Substrings")
    assert count_palindromic_substrings("abc") == 3
    assert count_palindromic_substrings("aaa") == 6
    print("   ✓ Palindromic Substrings tests passed")
    
    # Test Min Stack
    print("\n7. Min Stack")
    min_stack = MinStack()
    min_stack.push(-2)
    min_stack.push(0)
    min_stack.push(-3)
    assert min_stack.get_min() == -3
    min_stack.pop()
    assert min_stack.top() == 0
    assert min_stack.get_min() == -2
    print("   ✓ Min Stack tests passed")
    
    # Test Number of Islands
    print("\n8. Number of Islands")
    grid1 = [
        ["1", "1", "1", "1", "0"],
        ["1", "1", "0", "1", "0"],
        ["1", "1", "0", "0", "0"],
        ["0", "0", "0", "0", "0"]
    ]
    assert num_islands(grid1) == 1
    
    grid2 = [
        ["1", "1", "0", "0", "0"],
        ["1", "1", "0", "0", "0"],
        ["0", "0", "1", "0", "0"],
        ["0", "0", "0", "1", "1"]
    ]
    assert num_islands(grid2) == 3
    print("   ✓ Number of Islands tests passed")
    
    # Test Level Order
    print("\n9. Level Order Traversal")
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20, TreeNode(15), TreeNode(7))
    assert level_order(root) == [[3], [9, 20], [15, 7]]
    print("   ✓ Level Order tests passed")
    
    # Test Max Depth
    print("\n10. Max Depth")
    assert max_depth(root) == 3
    print("   ✓ Max Depth tests passed")
    
    # Test Valid BST
    print("\n11. Validate BST")
    bst_root = TreeNode(2)
    bst_root.left = TreeNode(1)
    bst_root.right = TreeNode(3)
    assert is_valid_bst(bst_root) == True
    print("   ✓ Validate BST tests passed")
    
    # Test BFS on Graph
    print("\n12. Graph BFS")
    graph = {
        0: [1, 2],
        1: [2],
        2: [0, 3],
        3: [3]
    }
    bfs_result = graph_bfs(graph, 2)
    assert bfs_result == [2, 0, 3, 1]
    print("   ✓ Graph BFS tests passed")
    
    # Test Topological Sort
    print("\n13. Topological Sort")
    result = topological_sort(6, [[3, 0], [3, 1], [4, 1], [4, 2], [5, 3], [5, 4]])
    assert len(result) == 6
    print("   ✓ Topological Sort tests passed")
    
    # Test Dijkstra
    print("\n14. Dijkstra's Algorithm")
    graph = {
        0: [(1, 4), (2, 1)],
        1: [(3, 1)],
        2: [(1, 2), (3, 5)],
        3: []
    }
    distances = dijkstra(graph, 0)
    assert distances[0] == 0
    assert distances[1] == 3
    assert distances[2] == 1
    assert distances[3] == 4
    print("   ✓ Dijkstra's tests passed")
    
    print("\n" + "=" * 80)
    print("✅ ALL TESTS PASSED!")
    print("=" * 80)
    print("\nKey Patterns Reviewed:")
    print("  1. Arrays & Hashing: Two Sum, 3Sum, Container, Group Anagrams")
    print("  2. String Manipulation: Longest Substring, Palindromic, strStr")
    print("  3. Stack & Queue: Min Stack, Eval RPN, Daily Temperatures, Islands")
    print("  4. Linked List: Merge, Reverse, Palindrome")
    print("  5. Trees: Level Order, Max Depth, BST, Diameter, Balanced")
    print("  6. Graphs: BFS, Topological Sort, Dijkstra")
    print("\nReady for advanced problem solving!")


if __name__ == "__main__":
    test_all()
